src/extensions/idle.rs

```diff
@@ -196,73 +191,55 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
-    /// Set the keep-alive interval to use when `wait_keepalive_while` is called.
+    /// Set the timeout duration on the connection. This will also set the frequency
+    /// at which the connection is refreshed.
+    ///
+    /// The interval defaults to 29 minutes as given in RFC 2177.
+    pub fn set_timeout(&mut self, interval: Duration) -> &mut Self {
+        self.timeout = interval;
+        self
+    }
+
+    /// Do not continuously refresh the IDLE connection in the background.
     ///
-    /// The interval defaults to 29 minutes as dictated by RFC 2177.
-    pub fn set_keepalive(&mut self, interval: Duration) {
-        self.keepalive = interval;
+    /// By default, connections will periodically be refreshed in the background using the
+    /// timeout duration set by [`set_timeout`]. If you do not want this behaviour, call
+    /// this function and the connection will simply IDLE until `wait_while` returns or
+    /// the timeout expires.
+    pub fn no_keepalive(&mut self) -> &mut Self {
+        self.keepalive = false;
+        self
     }
 
     /// Block until the given callback returns `false`, or until a response
     /// arrives that is not explicitly handled by [`UnsolicitedResponse`].
-    ///
-    /// This method differs from [`Handle::wait_while`] in that it will periodically refresh the IDLE
-    /// connection, to prevent the server from timing out our connection. The keepalive interval is
-    /// set to 29 minutes by default, as dictated by RFC 2177, but can be changed using
-    /// [`Handle::set_keepalive`].
-    ///
-    /// This is the recommended method to use for waiting.
-    pub fn wait_keepalive_while<F>(self, callback: F) -> Result<()>
+    pub fn wait_while<F>(&mut self, callback: F) -> Result<()>
     where
         F: FnMut(UnsolicitedResponse) -> bool,
     {
+        self.init()?;
         // The server MAY consider a client inactive if it has an IDLE command
         // running, and if such a server has an inactivity timeout it MAY log
         // the client off implicitly at the end of its timeout period.  Because
         // of that, clients using IDLE are advised to terminate the IDLE and
         // re-issue it at least every 29 minutes to avoid being logged off.
         // This still allows a client to receive immediate mailbox updates even
         // though it need only "poll" at half hour intervals.
-        let keepalive = self.keepalive;
-        self.timed_wait(keepalive, true, callback).map(|_| ())
-    }
-
-    /// Block until the given given amount of time has elapsed, the given callback
-    /// returns `false`, or until a response arrives that is not explicitly handled
-    /// by [`UnsolicitedResponse`].
-    pub fn wait_with_timeout_while<F>(self, timeout: Duration, callback: F) -> Result<WaitOutcome>
-    where
-        F: FnMut(UnsolicitedResponse) -> bool,
-    {
-        self.timed_wait(timeout, false, callback)
+        self.timed_wait(callback).map(|_| ())
```

Comment:
At this point is there even a point in having `timed_wait` be its own method?

Reply:
Not really. Do we want to propagate the `WaitOutcome` back to the caller since it's right there anyway?